File: 2010_2011_mf_sujet_4_convergences_de_processus/TPconv2_MF.m
%%Chaine d'Ehrenfest
X=zeros(1,500);
X(1)=1;
for n=2:500
    X(n)=Ehrenfest3(5,X(n-1));
end
plot(X);
%%
Ehr900=zeros(1,5000);
Ehr1000=zeros(1,5000);
for k=1:5000
    C=1;
    for n=2:900 
        C=Ehrenfest3(5,C);
    end
    Ehr900(k)=C;
    for n=901:1000
    C=Ehrenfest3(5,C);
    end
    Ehr1000(k)=C;
end
plotrep(Ehr900);
hold on
plotrep(Ehr1000);
stairs([0:5],pbinom([0:5],5,0.5),'r');
hold off

%% Processus de Poisson
X=zeros(1,500);
X=rexp(500,1);
stairs(cumsum(X),[1:500]);%Dessin de Nt en fonction de t 
                            %(en fait Tn en abscisse et N_Tn en ordonnee)
plot(cumsum(X),[1:500]./cumsum(X)','r'); %Nt/t en fonction de t
hold on;
plot(ones(1,500));%la limite theorique
hold off;
Ic=[500/sum(X)-qnorm(0.975)*sqrt(500)/sum(X),500/sum(X)+qnorm(0.975)*sqrt(500)/sum(X)]

%% Minimum de lois uniformes
X=rand(1,500);
m=zeros(1,500);
m(1)=X(1);
for k=2:500
    if X(k)<m(k-1) m(k)=X(k);
    else m(k)=m(k-1);
    end;
end;
plot(m);% convergence ps vers 0
m=zeros(1,5000);
for k=1:5000
    X=rand(1,500);
    m(k)=500*min(X);
end;
plotrep(m);%conv en loi vers une loi exp.
hold on
fplot(@(x) 1-exp(-x),[0 3],'r');
hold off

%% File d'attente
X=zeros(1,1000);
X(1)=0;
for k=2:1000
    X(k)=Fattente(X(k-1),0.1,0.6);
end;
plot(X,'o');
hold on;
X(1)=0;
for k=2:1000
    X(k)=Fattente(X(k-1),0.5,0.5);
end;
plot(X,'r+');
X(1)=0;
for k=2:1000
    X(k)=Fattente(X(k-1),0.6,0.1);
end;
plot(X,'g*');
hold off;
%% 
plotergo(0.1,0.6,10000);%la chaine semble bien ergodique
plotergo(0.5,0.55,10000);
%%
convattente(0.1,0.6) %il y a bien convergence en loi vers la loi pi
convattente(0.5,0.5) % on n'observe pas de convergence en loi
convattente(0.6,0.1) % on n'observe pas de convergence en loi

File: 2010_2011_mf_sujet_4_convergences_de_processus/TPconv2_MF.py
"""TP Convergences de processus — Master MF 2010-2011."""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import binom as binom_dist, norm

from Ehrenfest3 import Ehrenfest3
from Fattente import Fattente
from plotergo import plotergo
from convattente import convattente
from plotrep import plotrep

# %% Ehrenfest chain
X = np.zeros(500, dtype=int); X[0] = 1
for n in range(1, 500):
    X[n] = int(Ehrenfest3(5, X[n-1]))
plt.figure(); plt.plot(X); plt.title('Ehrenfest 500 steps'); plt.show()

# %% Convergence to binomial stationary distribution
Ehr900 = np.zeros(5000, dtype=int)
Ehr1000 = np.zeros(5000, dtype=int)
for k in range(5000):
    C = 1
    for n in range(1, 900):
        C = int(Ehrenfest3(5, C))
    Ehr900[k] = C
    for n in range(900, 1000):
        C = int(Ehrenfest3(5, C))
    Ehr1000[k] = C

plt.figure()
plotrep(Ehr900)
plotrep(Ehr1000)
k_arr = np.arange(6)
plt.step(k_arr, binom_dist.cdf(k_arr, 5, 0.5), 'r', where='post')
plt.show()

# %% Poisson process (exponential inter-arrivals)
inter = np.random.exponential(1, 500)
T = np.cumsum(inter)
plt.figure()
plt.step(T, np.arange(1, 501), where='post', label='N_t')
plt.plot(T, np.arange(1, 501) / T, 'r', label='N_t / t')
plt.axhline(1, color='k')
plt.legend(); plt.title('Poisson process'); plt.show()
Ic = np.array([500/np.sum(inter) - norm.ppf(0.975)*np.sqrt(500)/np.sum(inter),
               500/np.sum(inter) + norm.ppf(0.975)*np.sqrt(500)/np.sum(inter)])
print(f'IC taux: {Ic}')

# %% Minimum of uniforms → exponential
X = np.random.rand(500)
m_arr = np.zeros(500)
m_arr[0] = X[0]
for k in range(1, 500):
    m_arr[k] = min(X[k], m_arr[k-1])
plt.figure(); plt.plot(m_arr); plt.title('Running minimum → 0'); plt.show()

m_scaled = np.zeros(5000)
for k in range(5000):
    m_scaled[k] = 500 * np.min(np.random.rand(500))
plt.figure()
plotrep(m_scaled)
x = np.linspace(0, 3, 200)
plt.plot(x, 1 - np.exp(-x), 'r')
plt.title('500 * min(U_1,...,U_500) → Exp(1)'); plt.show()

# %% Queuing chain
X = np.zeros(1000, dtype=int)
plt.figure()
for (p, q, style) in [(0.1, 0.6, 'o'), (0.5, 0.5, 'r+'), (0.6, 0.1, 'g*')]:
    X[0] = 0
    for k in range(1, 1000):
        X[k] = Fattente(X[k-1], p, q)
    plt.plot(X, style, ms=3)
plt.title('File d\'attente'); plt.show()

# %% Ergodicity
plt.figure(); plotergo(0.1, 0.6, 10000); plt.suptitle('p=0.1, q=0.6 ergodique'); plt.show()
plt.figure(); plotergo(0.5, 0.55, 10000); plt.suptitle('p=0.5, q=0.55'); plt.show()

# %% Convergence in distribution
plt.figure(); convattente(0.1, 0.6); plt.title('p<q : converge'); plt.show()
plt.figure(); convattente(0.5, 0.5); plt.title('p=q : pas de convergence'); plt.show()
plt.figure(); convattente(0.6, 0.1); plt.title('p>q : pas de convergence'); plt.show()
